### PR TITLE
Feature focus timer

### LINKED ISSUE #150: [User Story Feature] Focus Timer
**User Story**
What is the user story this feature aims to achieve?  
Describe it and label the issue accordingly.  
As a easily distracted computing student, I want to use an offline focus tracker when I’m coding so that I can avoid distractions online and track my progress easily. 

**Is your feature request related to a problem? Please describe.**
When I use online focus timers through the internet, I often end up surfing the web which stops my focus and causes me to not be productive. 

**Describe the solution you'd like**
The focus timer should be able to countdown in cycles and, users should be able to indicated their length or work and length of break, length of long break and how many cycles they wish to complete.

**Additional context**
We will add on additional features to improve on it.

### PR DESCRIPTION
**Pull request by YongbinWang**
Implemented basic Focus Timer. This fixes #150.
**Current features includes:**

- Start - Start the session and used to start subsequent countdown timers.
- Pause- Pause the current countdown and print time paused.
- Resume - Resume current countdown and print time resumed.
- Stop - Stop the whole session and terminate timer thread.
- Check - Used to check current time left in the countdown.

**Suggestions for improvements/Features not yet implemented:**
Config Command - For users to configure work/break/long break/cycle accordingly else default will be used.
Restart Session Command - Current implementations returns countdownindex to first countdown. Take note all countdowns will have mins= 0  and seconds = 0 if session is completed.
Differentiate countdowns -  Either through extending Countdown Class or storing a type inside Countdown Class
Improve UI output - Better UI output for users.
Improve defensiveness of focus timer - heavy bug testing.
Add rough completion time when timer starts - Can be implemented using get current timestamp and adding time duration to it.
Countdown during last 10 seconds.
**Below is a sample output**
image

### CODE REVIEW

File: src/main/java/wellnus/focus/command/CheckCommand.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    private static final String CHECK_OUTPUT = "Time left: ";
+    private final Session session;
+    private final TextUi textUi;
+
+    /**
+     * Constructs a CheckCommand object.
+     * Session in FocusManager is passed into this class.
+     *
+     * @param arguments Argument-Payload Hashmap generated by CommandParser
+     * @param session   Session object which is an arraylist of Countdowns
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/command/CheckCommand.java:30]: Extra whitespace before `Session`

File: src/main/java/wellnus/focus/command/HomeCommand.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    private static final String NO_ADDITIONAL_MESSAGE = "";
+    private final TextUi textUi;
+    private final Session session;
+
+    /**
+     * Constructor for HomeCommand object.
+     * Session in FocusManager is passed into this class.
+     *
+     * @param arguments Argument-Payload map generated by CommandParser
+     * @param session   Session object which is an arraylist of Countdowns
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/command/HomeCommand.java:30]: Extra whitespace before `Session`

File: src/main/java/wellnus/focus/command/HomeCommand.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+     */
+    public HomeCommand(HashMap<String, String> arguments, Session session) {
+        super(arguments);
+        this.textUi = new TextUi();
+        this.session = session;
+    }
+
+    /**
+     * Check if a HomeCommand is executed and user wants to return to home
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/command/HomeCommand.java:39]: Missing '.' at the end of sentence

File: src/main/java/wellnus/focus/command/ResumeCommand.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    private static final String RESUME_OUTPUT = "Timer resumed at: ";
+    private final Session session;
+    private final TextUi textUi;
+
+    /**
+     * Constructs a ResumeCommand object.
+     * Session in FocusManager is passed into this class.
+     *
+     * @param arguments Argument-Payload Hashmap generated by CommandParser
+     * @param session   Session object which is an arraylist of Countdowns
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/command/ResumeCommand.java:30]: Extra whitespace before `Session`

File: src/main/java/wellnus/focus/command/StartCommand.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    private final Session session;
+
+    private final TextUi textUi;
+
+    /**
+     * Constructor for StartCommand object.
+     * Session in FocusManager is passed into this class.
+     *
+     * @param arguments Argument-Payload Hashmap generated by CommandParser
+     * @param session   Session object which is an arraylist of Countdowns
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/command/StartCommand.java:31]: Extra whitespace before `Session`

File: src/main/java/wellnus/focus/command/StartCommand.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+     * Identifies the feature that this Command is associated with. Override
+     * this in subclasses so toString() returns the correct String representation.
+     *
+     * @return String Keyword for the feature associated with this Command
+     */
+    @Override
+    protected String getFeatureKeyword() {
+        return FocusManager.FEATURE_NAME;
+    }
+
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/command/StartCommand.java:60]: Extra newline

File: src/main/java/wellnus/focus/command/StartCommand.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    public void execute() {
+        try {
+            validateCommand(super.getArguments());
+        } catch (BadCommandException badCommandException) {
+            textUi.printErrorFor(badCommandException, NO_ADDITIONAL_MESSAGE);
+            return;
+        }
+        session.checkPrevCountdown();
+        if (!session.getSession().get(session.getCurrentCountdownIndex()).getIsRunning()) {
+            if (session.getCurrentCountdownIndex() == 0) {
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/command/StartCommand.java:77]: Avoid magic literals

File: src/main/java/wellnus/focus/command/StopCommand.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+    private final Session session;
+    private final TextUi textUi;
+
+    /**
+     * Constructs a StopCommand object.
+     * Session in FocusManager is passed into this class.
+     *
+     * @param arguments Argument-Payload Hashmap generated by CommandParser
+     * @param session   Session object which is an arraylist of Countdowns
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/command/StopCommand.java:30]: Extra whitespace before `Session`

File: src/main/java/wellnus/focus/feature/Countdown.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    private int seconds;
+    private final String description;
+    private AtomicBoolean isCompletedCountdown;
+    private AtomicBoolean isRunClock;
+
+    /**
+     * Constructor of Countdown.
+     * This will initialise the private attributes of Countdown object.
+     *
+     * @param minutes     the number of minutes to countdown
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/Countdown.java:36]: Extra whitespace before `the`, and capitalise 't' of 'the'

File: src/main/java/wellnus/focus/feature/Countdown.java [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+     * @param description description of the current task user is focusing on
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/Countdown.java:37]: Capitalise 'd' of 'description'

File: src/main/java/wellnus/focus/feature/Countdown.java [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+     * This method will pause the countdown timer.
+     */
+    public void setPause() {
+        isRunClock.set(false);
+    }
+
+    /**
+     * This method will return the status of the countdown timer.
+     *
+     * @return true if the countdown timer is running, false otherwise
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/Countdown.java:113]: Capitalise 'i' of 'if'

File: src/main/java/wellnus/focus/feature/Countdown.java [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+     */
+    public boolean getIsRunning() {
+        return isRunClock.get();
+    }
+
+    /**
+     * This method will return the current minutes of the countdown timer.
+     *
+     * @return the minutes of the countdown timer
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/Countdown.java:122]: Capitalise 't' of 'the'

File: src/main/java/wellnus/focus/feature/Countdown.java [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+     */
+    public int getMinutes() {
+        return this.minutes;
+    }
+
+    /**
+     * This method will return the current seconds of the countdown timer.
+     *
+     * @return the seconds of the countdown timer
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/Countdown.java:131]: Capitalise 't' of 'the'

File: src/main/java/wellnus/focus/feature/Countdown.java [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+     */
+    public int getSeconds() {
+        return this.seconds;
+    }
+
+    /**
+     * This method will return the description of the countdown timer.
+     *
+     * @return the description of the countdown timer
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/Countdown.java:140]: Capitalise 't' of 'the'

File: src/main/java/wellnus/focus/feature/FocusManager.java [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+            return new CheckCommand(arguments, session);
+        default:
+            throw new BadCommandException(UNKNOWN_COMMAND_MESSAGE);
+        }
+    }
+
+    private void greet() {
+        textUi.printOutputMessage(FOCUS_TIMER_GREET);
+    }
+
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/FocusManager.java:84]: Extra newline

File: src/main/java/wellnus/focus/feature/FocusManager.java [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+            } catch (WellNusException exception) {
+                textUi.printErrorFor(exception, "Check user guide for valid commands!");
+            }
+        }
+    }
+
+    /**
+     * Utility function to get the featureName this Manager is administering
+     *
+     * @return name of the feature that this Manager handles
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/FocusManager.java:106]: Capitalise 'n' of 'name'

File: src/main/java/wellnus/focus/feature/FocusManager.java [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+     */
+    @Override
+    public String getFeatureName() {
+        return FEATURE_NAME;
+    }
+
+    /**
+     * Utility function to get a summary description of the feature this Manager is administering
+     *
+     * @return summary description of the feature that this Manager handles
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/FocusManager.java:116]: Capitalise 's' of 'summary'

File: src/main/java/wellnus/focus/feature/FocusManager.java [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+     */
+    @Override
+    public String getBriefDescription() {
+        return FEATURE_BRIEF_DESCRIPTION;
+    }
+
+    /**
+     * Utility function to get the full description of the feature this Manager is administering
+     *
+     * @return full description of the feature that this Manager handles
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/FocusManager.java:126]: Capitalise 'f' of 'full'

File: src/main/java/wellnus/focus/feature/FocusManager.java [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+     *
+     * @return summary description of the feature that this Manager handles
+     */
+    @Override
+    public String getBriefDescription() {
+        return FEATURE_BRIEF_DESCRIPTION;
+    }
+
+    /**
+     * Utility function to get the full description of the feature this Manager is administering
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/FocusManager.java:124]: Add '.' at the end of the sentence

File: src/main/java/wellnus/focus/feature/FocusManager.java [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+                String noAdditionalMessage = "";
+                textUi.printErrorFor(badCommandException, noAdditionalMessage);
+            } catch (WellNusException exception) {
+                textUi.printErrorFor(exception, "Check user guide for valid commands!");
+            }
+        }
+    }
+
+    /**
+     * Utility function to get the featureName this Manager is administering
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/FocusManager.java:104]: Add '.' at the end of the sentence

File: src/main/java/wellnus/focus/feature/FocusManager.java [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+     *
+     * @return name of the feature that this Manager handles
+     */
+    @Override
+    public String getFeatureName() {
+        return FEATURE_NAME;
+    }
+
+    /**
+     * Utility function to get a summary description of the feature this Manager is administering
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/FocusManager.java:114]: Add '.' at the end of the sentence

File: src/main/java/wellnus/focus/feature/FocusManager.java [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+     *
+     * @return full description of the feature that this Manager handles
+     */
+    @Override
+    public String getFullDescription() {
+        return FEATURE_FULL_DESCRIPTION;
+    }
+
+    /**
+     * Utility function to set a list of main commands the feature supports <br>
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/FocusManager.java:134]: Add '.' at the end of the sentence

File: src/main/java/wellnus/focus/feature/FocusManager.java [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+    @Override
+    public void runEventDriver() throws BadCommandException {
+        greet();
+        runCommands();
+    }
+
+    /**
+     * Public method used for testing FocusManager's handling of invalidCommands.
+     *
+     * @param userCommand simulated user input
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/FocusManager.java:156]: Capitalise 's' of 'simulated'

File: src/main/java/wellnus/focus/feature/Session.java [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        Countdown longBreak = new Countdown(longBrk, "Long Break");
+        session.remove((cycle * 2) - 1);
+        session.add(longBreak);
+    }
+
+    /**
+     * Method to get the session.
+     *
+     * @return arraylist of countdown objects
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/Session.java:46]: Capitalise 'a' of 'arraylist'

File: src/main/java/wellnus/focus/feature/Session.java [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+     */
+    public ArrayList<Countdown> getSession() {
+        return this.session;
+    }
+
+    /**
+     * Method to get the current countdown index.
+     *
+     * @return the current countdown index
```
Comment (haoyangw) [inline src/main/java/wellnus/focus/feature/Session.java:55]: Capitalise 't' of 'the'

File: src/test/java/wellnus/focus/FocusTest.java [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+    public void setUp() {
+        System.setOut(new PrintStream(outputStreamCaptor));
+    }
+
+
+    /**
+     * Test whether start command starts the timer program.
+     * Stops program immediately after.
+     *
+     * @throws BadCommandException
```
Comment (haoyangw) [inline src/test/java/wellnus/focus/FocusTest.java:62]: Remember to fill in the description for this `throws`!

File: src/test/java/wellnus/focus/FocusTest.java [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+     */
+    @Test
+    void startTimer_checkResult_success() throws WellNusException {
+        HashMap<String, String> arguments = parser.parseUserInput(START_COMMAND);
+        StartCommand startCommand = new StartCommand(arguments, session);
+        startCommand.execute();
+        Assertions.assertEquals(true, session.getSession().get(0).getIsRunning());
+        session.getSession().get(0).setStop();
```
Comment (haoyangw) [inline src/test/java/wellnus/focus/FocusTest.java:70]: Avoid magic literals!

File: src/test/java/wellnus/focus/FocusTest.java [hunk 28]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    /**
+     * Test FocusManager to throw {@link BadCommandException} when an invalid command is given to the FocusManager
```
Comment (haoyangw) [inline src/test/java/wellnus/focus/FocusTest.java:74]: Add '.' at the end of the sentence

File: src/test/java/wellnus/focus/FocusTest.java [hunk 29]
```diff
@@ -0,0 +1,108 @@
...
+        // Test false command by user
+        FocusManager focusManager = new FocusManager();
+        String command = "wrong";
+        Assertions.assertThrows(BadCommandException.class, () -> {
+            focusManager.testInvalidCommand(command);
+        });
+    }
+
+    /**
+     * Test if session creates countdowns
```
Comment (haoyangw) [inline src/test/java/wellnus/focus/FocusTest.java:87]: Add '.' at the end of the sentence

File: src/test/java/wellnus/focus/FocusTest.java [hunk 30]
```diff
@@ -0,0 +1,108 @@
...
+    /**
+     * Test if CheckCommand provides intended output
+     */
+    @Test
+    public void checkCommand_checkOutput_success() throws WellNusException {
+        HashMap<String, String> arguments = parser.parseUserInput(CHECK_COMMAND);
+        CheckCommand checkCommand = new CheckCommand(arguments, session);
+        checkCommand.execute();
+        String expectedCheckOutput = "Time left: " + session.getSession().get(0).getMinutes()
+                + ":" + session.getSession().get(0).getSeconds();
```
Comment (haoyangw) [inline src/test/java/wellnus/focus/FocusTest.java:103]: Avoid magic literals!

File: src/test/java/wellnus/focus/FocusTest.java [hunk 31]
```diff
@@ -0,0 +1,108 @@
...
+    /**
+     * Test if session creates countdowns
+     */
+    @Test
+    public void session_createCountdowns_success() {
+        Assertions.assertEquals(true, session.getSession().size() > 0);
+    }
+
+    /**
+     * Test if CheckCommand provides intended output
```
Comment (haoyangw) [inline src/test/java/wellnus/focus/FocusTest.java:95]: Add '.' at the end of the sentence

**Review by haoyangw — CHANGES_REQUESTED** [review]
Hmm can't test your feature out right now, remember to add `this.getSupportedFeatureManagers().add(new FocusManager());` to `MainManager.setSupportedFeatureManagers()`!

**Comment by haoyangw** [comment]
Hmm potential bug if you call `check` after calling `stop`:
image
In fact, the timer is still saved at the countdown time(is this intended behaviour) even after quitting `FocusManager` and coming back to `FocusManager`, and if you then call `start`, the app crashes:
image
`try..catch` block for `IllegalStateException` will be necessary.

File: src/main/java/wellnus/focus/command/HomeCommand.java [hunk 32]
```diff
@@ -0,0 +1,107 @@
...
+        if (arguments.get(COMMAND_KEYWORD) != "") {
+            throw new BadCommandException(HomeCommand.COMMAND_INVALID_ARGUMENTS_MESSAGE);
+        }
+        if (!arguments.containsKey(HomeCommand.COMMAND_KEYWORD)) {
+            throw new BadCommandException(HomeCommand.COMMAND_INVALID_COMMAND_MESSAGE);
+        }
+    }
+}
+
+
```
Comment (wenxin-c) [inline src/main/java/wellnus/focus/command/HomeCommand.java:107]: Leave one new line at the end of the file is sufficient

File: src/main/java/wellnus/focus/feature/Countdown.java [hunk 33]
```diff
@@ -0,0 +1,146 @@
+package wellnus.focus.feature;
+
+
```
Comment (wenxin-c) [inline src/main/java/wellnus/focus/feature/Countdown.java:3]: Extra line here, line 76, 145

File: src/main/java/wellnus/focus/feature/FocusManager.java [hunk 34]
```diff
@@ -0,0 +1,193 @@
...
+    private static final String RESUME_COMMAND_KEYWORD = "resume";
+    private static final String CONFIG_COMMAND_KEYWORD = "config";
+    private static final String HOME_COMMAND_KEYWORD = "home";
+    private static final String STOP_COMMAND_KEYWORD = "stop";
+    private static final String CHECK_COMMAND_KEYWORD = "check";
+    private static final String UNKNOWN_COMMAND_MESSAGE = "No such command in focus timer!";
+    private static final String FOCUS_TIMER_GREET = "Welcome to Focus Timer";
+    private static final String COMMAND_KEYWORD_ASSERTION = "The key cannot be null"
+            + ", check user-guide for valid commands";
+
```
Comment (wenxin-c) [inline src/main/java/wellnus/focus/feature/FocusManager.java:37]: Extra line here

**Review by wenxin-c — COMMENTED** [review]
Generally the functionality has been achieved, there are many extra lines, change the styling error should be good.

File: src/main/java/wellnus/focus/feature/Session.java [hunk 35]
```diff
@@ -0,0 +1,79 @@
+package wellnus.focus.feature;
+
+
+import java.util.ArrayList;
+
+/**
+ * Represents a session of Countdown objects.
+ * A session is a sequence of Countdown objects represented by an ArrayList.
+ */
+public class Session {
```
Comment (nichyjt) [inline src/main/java/wellnus/focus/feature/Session.java:10]: As per one of my comments, add in a method to reset the state of the timer to allow for consecutive Sessions to be made. Alternatively, destroy the session object and create a new instance.

**Review by nichyjt — CHANGES_REQUESTED** [review]
Thank you @YongbinWang for the hard work. This was definitely not an easy feature to implement and I applaud the effort made to implement it in a pretty elegant way. 💪🏼 

There are 1-2 major issues with the implementation thus far which I would like you to consider. This is especially important as we are delving into more performance-sensitive logic (timer).

Please take a look at my attached comments and consider them 👍🏼

File: src/main/java/wellnus/focus/feature/FocusManager.java [hunk 36]
```diff
@@ -0,0 +1,193 @@
...
+    private final Session session;
+
+    /**
+     * Constructs a FocusManager object.
+     * Initialise a session and textUi.
+     * Session will be passed into different commands to be utilised.
+     */
+    public FocusManager() {
+        this.textUi = new TextUi();
+        this.session = new Session();
```
[the remark this answers is not in the document]
Comment (YongbinWang) [reply src/main/java/wellnus/focus/feature/FocusManager.java:49]: Made the code more defensive. In the latest commit.

File: src/main/java/wellnus/focus/feature/Session.java (hunk 35 above)
Comment (YongbinWang) [reply src/main/java/wellnus/focus/feature/Session.java:10]: Will leave it to be implemented, current method of doing so is not foolproof.

File: src/main/java/wellnus/focus/feature/Countdown.java [hunk 37]
```diff
@@ -0,0 +1,146 @@
...
+                        isCompletedCountdown.set(true);
+                        timer.cancel();
+                        java.awt.Toolkit.getDefaultToolkit().beep();
+                        textUi.printOutputMessage("Type start to begin the next countdown");
+                    } else if (seconds == DEFAULT_STOP_TIME) {
+                        seconds = DEFAULT_SECONDS;
+                        minutes--;
+                    } else {
+                        seconds--;
+                    }
```
[the remark this answers is not in the document]
Comment (YongbinWang) [reply src/main/java/wellnus/focus/feature/Countdown.java]: Can be changed accordingly, implemented it using a straightforward method.

File: src/main/java/wellnus/focus/feature/FocusManager.java [hunk 38]
```diff
@@ -0,0 +1,193 @@
...
+            try {
+                String commandString = textUi.getCommand();
+                Command command = getCommandFor(commandString);
+                command.execute();
+                isExit = HomeCommand.isExit(command);
+            } catch (BadCommandException badCommandException) {
+                String noAdditionalMessage = "";
+                textUi.printErrorFor(badCommandException, noAdditionalMessage);
+            } catch (WellNusException exception) {
+                textUi.printErrorFor(exception, "Check user guide for valid commands!");
```
[the remark this answers is not in the document]
Comment (YongbinWang) [reply src/main/java/wellnus/focus/feature/FocusManager.java]: Addressed.

File: src/main/java/wellnus/focus/feature/Session.java [hunk 39]
```diff
@@ -0,0 +1,79 @@
...
+     */
+    private void fillSession() {
+        for (int i = 0; i < cycle; i++) {
+            Countdown workCountDown = new Countdown(work, "Task");
+            Countdown breakCountDown = new Countdown(brk, "Break");
+            session.add(workCountDown);
+            session.add(breakCountDown);
+        }
+        Countdown longBreak = new Countdown(longBrk, "Long Break");
+        session.remove((cycle * 2) - 1);
```
[the remark this answers is not in the document]
Comment (YongbinWang) [reply src/main/java/wellnus/focus/feature/Session.java]: Thanks for the tip, addressed.

File: src/main/java/wellnus/focus/feature/FocusManager.java [hunk 40]
```diff
@@ -0,0 +1,193 @@
...
+     */
+    @Override
+    public void runEventDriver() throws BadCommandException {
+        greet();
+        runCommands();
+    }
+
+    /**
+     * Public method used for testing FocusManager's handling of invalidCommands.
+     *
```
[the remark this answers is not in the document]
Comment (YongbinWang) [reply src/main/java/wellnus/focus/feature/FocusManager.java:142]: Had issues testing commands, so I utilised Haoyang's method from his improvements to atomichabits. Open to suggestions.

File: src/main/java/wellnus/focus/feature/Countdown.java [hunk 41]
```diff
@@ -0,0 +1,146 @@
...
+        this.description = description;
+        this.textUi = new TextUi();
+    }
+
+    /**
+     * This method will start the countdown timer.
+     * Timer will notify user when the countdown timer has completed by playing a beep sound.
+     * The timer will run in the background and will be stopped when the countdown minutes and seconds reaches 0.
+     */
+    public void start() {
```
[the remark this answers is not in the document]
Comment (YongbinWang) [reply src/main/java/wellnus/focus/feature/Countdown.java:84]: Addressed.

File: src/main/java/wellnus/focus/feature/Countdown.java [hunk 42]
```diff
@@ -0,0 +1,146 @@
...
+        assert isRunClock.get() == false : TIMER_NOT_RUNNING_ASSERTION;
+        timer.scheduleAtFixedRate(new TimerTask() {
+            @Override
+            public void run() {
+                if (isRunClock.get()) {
+                    if (seconds == DEFAULT_STOP_TIME && minutes == DEFAULT_STOP_TIME) {
+                        isCompletedCountdown.set(true);
+                        timer.cancel();
+                        java.awt.Toolkit.getDefaultToolkit().beep();
+                        textUi.printOutputMessage("Type start to begin the next countdown");
```
[the remark this answers is not in the document]
Comment (YongbinWang) [reply src/main/java/wellnus/focus/feature/Countdown.java]: Addressed.

File: src/main/java/wellnus/focus/feature/Session.java [hunk 43]
```diff
@@ -0,0 +1,79 @@
...
+     * Method to fill the session with Countdown objects in a specific sequence.
+     */
+    private void fillSession() {
+        for (int i = 0; i < cycle; i++) {
+            Countdown workCountDown = new Countdown(work, "Task");
+            Countdown breakCountDown = new Countdown(brk, "Break");
+            session.add(workCountDown);
+            session.add(breakCountDown);
+        }
+        Countdown longBreak = new Countdown(longBrk, "Long Break");
```
[the remark this answers is not in the document]
Comment (YongbinWang) [reply src/main/java/wellnus/focus/feature/Session.java]: Addressed.

File: src/main/java/wellnus/focus/feature/Session.java [hunk 44]
```diff
@@ -0,0 +1,79 @@
...
+        fillSession();
+    }
+
+    /**
+     * Method to fill the session with Countdown objects in a specific sequence.
+     */
+    private void fillSession() {
+        for (int i = 0; i < cycle; i++) {
+            Countdown workCountDown = new Countdown(work, "Task");
+            Countdown breakCountDown = new Countdown(brk, "Break");
```
[the remark this answers is not in the document]
Comment (YongbinWang) [reply src/main/java/wellnus/focus/feature/Session.java]: Addressed.

File: src/main/java/wellnus/focus/command/CheckCommand.java [hunk 45]
```diff
@@ -0,0 +1,98 @@
...
+     * toString() returns the correct String representation.
+     *
+     * @return String Keyword of this Command
+     */
+    @Override
+    protected String getCommandKeyword() {
+        return COMMAND_KEYWORD;
+    }
+
+
```
[the remark this answers is not in the document]
Comment (YongbinWang) [reply src/main/java/wellnus/focus/command/CheckCommand.java]: Tried to fix most of the missing lines however my IDE quite troublesome sometimes. Will double check again when working on this feature. This comment applies to rest of the comments below.

File: src/main/java/wellnus/focus/command/CheckCommand.java (hunk 1 above)
Comment (YongbinWang) [reply src/main/java/wellnus/focus/command/CheckCommand.java:29]: IDE auto formatter aligns the start of the Description, not sure if this is a concern.

File: src/main/java/wellnus/focus/command/HomeCommand.java (hunk 3 above)
Comment (YongbinWang) [reply src/main/java/wellnus/focus/command/HomeCommand.java]: Addressed.

**Review by nichyjt — APPROVED** [review]

**Review by haoyangw — APPROVED** [review]
Some very minor code style violations and known bugs that are already listed in the PR description. Not ideal, but realistically, this is sufficient. Room for enhancement PRs!

**Pull request merged**